dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -552,39 +529,52 @@ void FramesDecoder::SeekFrame(int frame_id) {
   //  - for CFR, when we know pts, but don't know keyframes
 
   DALI_ENFORCE(
-    frame_id >= 0 && frame_id < NumFrames(),
-    make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
+      frame_id >= 0 && frame_id < NumFrames(),
+      make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
 
-  auto &frame_entry = Index(frame_id);
-  int keyframe_id = frame_entry.last_keyframe_id;
-  auto &keyframe_entry = Index(keyframe_id);
+  if (frame_id == next_frame_idx_) {
+    return;  // No need to seek
+  }
 
-  LOG_LINE << "Seeking to frame " << frame_id << " timestamp " << frame_entry.pts << std::endl;
+  // If we have an index, we can seek to the nearest keyframe first
+  if (HasIndex()) {
+    const auto &current_frame = Index(next_frame_idx_);
+    const auto &requested_frame = Index(frame_id);
+    auto keyframe_id = requested_frame.last_keyframe_id;
+    if (current_frame.last_keyframe_id != keyframe_id) {
+      // We are seeking to a different keyframe than the current frame,
+      // so we need to seek to the keyframe first
+      auto &keyframe_entry = Index(keyframe_id);
+      LOG_LINE << "Seeking to key frame " << keyframe_id << " timestamp " << keyframe_entry.pts 
+               << " for requested frame " << frame_id << " timestamp " << requested_frame.pts << std::endl;
+
+      // Seeking clears av buffers, so reset flush state info
+      if (flush_state_) {
+        while (ReadFlushFrame(nullptr, false)) {}
+        flush_state_ = false;
+      }
 
-  // Seeking clears av buffers, so reset flush state info
-  if (flush_state_) {
-    while (ReadFlushFrame(nullptr, false)) {}
-    flush_state_ = false;
+      int ret =
+        av_seek_frame(av_state_->ctx_, av_state_->stream_id_, keyframe_entry.pts, AVSEEK_FLAG_FRAME);
+      DALI_ENFORCE(ret >= 0,
+                make_string("Failed to seek to keyframe", keyframe_id,
+                            "in video \"", Filename(), "\" due to ", detail::av_error_string(ret)));
+      avcodec_flush_buffers(av_state_->codec_ctx_);
+      next_frame_idx_ = keyframe_id;
+    }
+  } else if (frame_id < next_frame_idx_) {
+    LOG_LINE << "Seeking to start of stream since requested frame " << frame_id << " is before current frame " << next_frame_idx_ << " and no index is available" << std::endl;
+    int ret =
+      av_seek_frame(av_state_->ctx_, av_state_->stream_id_, 0, AVSEEK_FLAG_BACKWARD);
+    DALI_ENFORCE(ret >= 0,
+                 make_string("Failed to seek to the first frame in video \"", Filename(),
+                             "\" due to ", detail::av_error_string(ret)));
+    avcodec_flush_buffers(av_state_->codec_ctx_);
+    next_frame_idx_ = 0;
```

Comment:
I think we can just call `Reset()`.

Reply:
Also there are files we cannot seek - see `ParseNumFrames` function. So in this case I guess we need to:
```
if (IsFormatSeekable()) {
    Reset();
} else {
    memory_video_file_->Seek(0, SEEK_SET);
    av_state_ = std::make_unique<AvState>();
}
```